File: 42.위상 정렬/3665.최종 순위/main.py
import sys

input = sys.stdin.readline

if __name__ == "__main__":
    T = int(input())

    for SUB_T in range(T):
        n = int(input())
        lst = [0] + list(map(int, input().split()))

        next_lst = [set() for _ in range(n + 1)]
        in_cnt = [0 for _ in range(n + 1)]
        before = set()
        for i in range(n, 0, -1):
            next_lst[lst[i]] = set(before)
            in_cnt[lst[i]] = i - 1
            before.add(lst[i])
        m = int(input())
        for _ in range(m):
            a, b = map(int, input().split())
            if b in next_lst[a]:
                next_lst[a].remove(b)
                in_cnt[b] -= 1
                next_lst[b].add(a)
                in_cnt[a] += 1
            else:
                next_lst[b].remove(a)
                in_cnt[a] -= 1
                next_lst[a].add(b)
                in_cnt[b] += 1

        stack = list()
        for i in range(1, n + 1):
            if in_cnt[i] == 0:
                stack.append(i)

        result = []
        while stack:
            if len(stack) > 1:
                break
            cur = stack.pop()
            result.append(str(cur))

            for next in next_lst[cur]:
                in_cnt[next] -= 1
                if in_cnt[next] == 0:
                    stack.append(next)
        if stack or len(result) != n:
            print("IMPOSSIBLE")
        else:
            print(" ".join(result))

File: 42.위상 정렬/3665.최종 순위/Main.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.Stack;
import java.util.StringTokenizer;

public class Main {
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;
        int T = Integer.parseInt(br.readLine());

        StringBuilder sb = new StringBuilder();

        while (T-- > 0) {
            int n = Integer.parseInt(br.readLine());
            int[] arr = new int[n];

            st = new StringTokenizer(br.readLine());
            for (int i = 0; i < arr.length; i++) {
                arr[i] = Integer.parseInt(st.nextToken());
            }

            List<Set<Integer>> next = new ArrayList<>();
            for (int i = 0; i <= arr.length; i++) {
                next.add(new HashSet<>());
            }
            int[] inDegree = new int[n + 1];

            Set<Integer> tempSet = new HashSet<>();
            int tempCnt = n - 1;
            for (int i = n - 1; i >= 0; i--) {
                inDegree[arr[i]] = tempCnt--;
                next.get(arr[i]).addAll(tempSet);
                tempSet.add(arr[i]);
            }

            int m = Integer.parseInt((br.readLine()));
            while (m-- > 0) {
                st = new StringTokenizer(br.readLine());
                int a = Integer.parseInt(st.nextToken());
                int b = Integer.parseInt(st.nextToken());

                if (next.get(b).contains(a)) {
                    // b->a -> a->b
                    next.get(b).remove(a);
                    inDegree[a] -= 1;
                    next.get(a).add(b);
                    inDegree[b] += 1;
                } else {
                    // a->b -> b->a
                    next.get(a).remove(b);
                    inDegree[b] -= 1;
                    next.get(b).add(a);
                    inDegree[a] += 1;
                }
            }

            Stack<Integer> stack = new Stack<>();
            for (int i = 1; i <= n; i++) {
                if (inDegree[i] == 0) {
                    stack.add(i);
                }
            }

            StringBuilder tempSb = new StringBuilder();
            int cnt = 0;
            while (!stack.isEmpty()) {
                if (stack.size() > 1) {
                    break;
                }

                int cur = stack.pop();
                cnt += 1;
                tempSb.append(cur).append(" ");

                for (int ne : next.get(cur)) {
                    inDegree[ne] -= 1;
                    if (inDegree[ne] == 0) {
                        stack.add(ne);
                    }
                }
            }

            if (cnt == n) {
                sb.append(tempSb).append("\n");
            } else {
                sb.append("IMPOSSIBLE").append("\n");
            }
        }

        System.out.println(sb.toString());
    }
}